### PR TITLE
Infer schema from gremlin/sparql query results for Neptune qpt

### PR DESCRIPTION
**Pull request by VenkatasivareddyTR**
*Issue #, if available:*

*Description of changes:*
Currently we are generating schema from aws glue for Neptune qpt as well
[Neptune_QPT_Tests.xlsx](https://github.com/user-attachments/files/18505470/Neptune_QPT_Tests.xlsx)
, this change will get schema from gremlin/sparql query results for Neptune qpt.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneMetadataHandler.java [hunk 1]
```diff
@@ -296,7 +291,7 @@ public GetTableResponse doGetQueryPassthroughSchema(BlockAllocator allocator, Ge
                 Client client = neptuneConnection.getNeptuneClientConnection();
                 GraphTraversalSource graphTraversalSource = neptuneConnection.getTraversalSource(client);
                 String gremlinQuery = qptArguments.get(NeptuneQueryPassthrough.QUERY);
-                gremlinQuery = gremlinQuery.concat(".limit(1)");
+                gremlinQuery = gremlinQuery.concat(".limit(10)");
```
Comment (chngpe) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneMetadataHandler.java:294]: Is it possible that the first 10 row data is different than the rest of rows?

Example:
Can first 10 rows contains a different result set which has no column field call `resource_type` and 11~20 rows has that? If yes, the schema inferred will drop the column? (and vice-versa, if no column in first 10 rows and data coming from 11~20, when reading data will there be an issue?)

It is okay, if not supported but i want to document the behavior as documentation

Thanks team

Comment (VenkatasivareddyTR) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneMetadataHandler.java:294]: Yes, we can update documentation about this limitation.

We determine the schema by examining the first 10 records. If the data exhibits schema variations in further rows, then we may not accurately capture the complete schema. It seems similar approach is followed in DynamoDB connector as well.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java [hunk 2]
```diff
@@ -96,61 +96,66 @@ public boolean hasNext()
...
-                    logger.debug("Int" + varName + "=" + oval);
-                } 
-                else if (type.equals(XMLSchema.INTEGER)) {
-                    oval = lit.integerValue();
-                    logger.debug("Integer " + varName + "=" + oval);
-                } 
-                else if (type.equals(XMLSchema.LONG)) {
-                    oval = lit.longValue();
-                    logger.debug("Long " + varName + "=" + oval);
+            if (val != null) {
```
Comment (AbdulR3hman) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java:106]: are we saying here that if value is null; we are still adding it to `ret` map?

Comment (AbdulR3hman) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java:106]: alternatively; if we want know that if `value == null` then we return as is; why not check for that and `continue;` 

that way the next code is not indent and easier to read.

Comment (VenkatasivareddyTR) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java:106]: > are we saying here that if value is null; we are still adding it to `ret` map?

Yes, to maintain consistency with SPARQL query results, which include column names with empty values, null values are also added to the 'ret' map.

Comment (VenkatasivareddyTR) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java:106]: > alternatively; if we want know that if `value == null` then we return as is; why not check for that and `continue;`
> 
> that way the next code is not indent and easier to read.

If we use `value == null` then it's not consistent with SPARQL query results.
image

SPARQL query results:
image

With existing PR code changes:
image

File: athena-neptune/src/test/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.ArrayList;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(MockitoJUnitRunner.class)
+public class NeptuneSchemaUtilsTest {
+    private static final Logger logger = LoggerFactory.getLogger(NeptuneSchemaUtilsTest.class);
+
+    private final Map<String, Object> objectMap = ImmutableMap.of(
```
Comment (AbdulR3hman) [inline athena-neptune/src/test/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtilsTest.java:40]: can we also simulate a null as a test

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java (hunk 2 above)
Comment (AbdulR3hman) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java:106]: understood; sounds good.

File: athena-neptune/src/test/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtilsTest.java (hunk 3 above)
Comment (VenkatasivareddyTR) [reply athena-neptune/src/test/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtilsTest.java]: Hi @AbdulR3hman, If I understand correctly, a null value needs to be added to the objectMap? If so, I have added a null value to objectMap. 
Alternatively, if you expect the objectMap itself to be null, that is not applicable here, as it is a NotNull parameter.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneQueryPassthrough.java [hunk 4]
```diff
@@ -56,7 +57,7 @@ public String getFunctionName()
     @Override
     public List<String> getFunctionArguments()
     {
-        return Arrays.asList(DATABASE, COLLECTION, QUERY);
+        return Arrays.asList(DATABASE, COLLECTION, QUERY, COMPONENT_TYPE);
```
Comment (AbdulR3hman) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneQueryPassthrough.java]: this class should be split into two classes; QPT can support multiple signatures; 

We should have different implementation for SPRAQL and Gremlin; one should be called `query` and the other `traverse`

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneMetadataHandler.java (hunk 1 above)
Comment (AbdulR3hman) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneMetadataHandler.java]: Please declare this as a static variable in `com.amazonaws.athena.connectors.neptune.Constants` class

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        SchemaBuilder schemaBuilder = SchemaBuilder.newBuilder();
+        //Building schema from gremlin/sparql query results.
+        resultsMap.forEach((columnName, columnValue) -> buildSchema(columnName.toString(), columnValue,  schemaBuilder));
+        schemaBuilder.addMetadata(Constants.SCHEMA_COMPONENT_TYPE, componentTypeValue);
+        schemaBuilder.addMetadata(Constants.SCHEMA_GLABEL, tableName);
+        schema = schemaBuilder.build();
+        return schema;
+    }
+
+    private static void buildSchema(String columnName, Object columnValue, SchemaBuilder schemaBuilder)
```
Comment (AbdulR3hman) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java:55]: We have two clients here; Gremlin and SPARQL; are we saying that the Neptune connection's result will return the same types regardless of the client we are using? based on what are we making this assumption? 

If not; we should have different schema builder and type conversion from Grelmin to arrow; and another for SPARQL?

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneQueryPassthrough.java (hunk 4 above)
Comment (VenkatasivareddyTR) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneQueryPassthrough.java]: Thank you @AbdulR3hman for reviewing, I have added code to support multiple signatures for neptune qpt.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java (hunk 5 above)
Comment (VenkatasivareddyTR) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java:55]: Hi @AbdulR3hman Yes, that's correct. 

For Gremlin queries, we are currently limiting support to valueMap() traversals. These traversals return Object types, which our connection logic explicitly converts into a Map representation.

For SPARQL queries, the direct response from Neptune is already structured as a Map.

Therefore, both Gremlin (after conversion) and SPARQL results converge to a Map type. This enables us to use a single, unified function for building the schema from these results.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneGremlinQueryPassthrough.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Override
+    public Logger getLogger()
+    {
+        return LOGGER;
+    }
+
+    @Override
+    public void customConnectorVerifications(Map<String, String> engineQptArguments)
```
Comment (fal-bharadwaj) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneGremlinQueryPassthrough.java:78]: I think we should add more validations based on the arguments like null checks and gremlin syntax validations?

File: athena-neptune/src/test/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(MockitoJUnitRunner.class)
+public class NeptuneSchemaUtilsTest {
```
Comment (fal-bharadwaj) [inline athena-neptune/src/test/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtilsTest.java]: The test can be better organized with a setup that does initialization, and separate test cases for positive and negative usecases including edge cases.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneGremlinQueryPassthrough.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        // Verify no mixed operations (SPARQL and Gremlin in same request)
+        if (engineQptArguments.containsKey(QUERY)) {
```
Comment (fal-bharadwaj) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneGremlinQueryPassthrough.java:85]: I think we should add more validations based on the arguments like null checks and gremlin syntax validations?

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java (hunk 5 above)
Comment (fal-bharadwaj) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java:55]: Just so I understand correctly, 

 > For Gremlin queries, we are currently limiting support to valueMap() traversals.

- Is that limitation for neptune, federated neptune connector or for QPT feature?

- even though both Gremlin and SPARQL return a Map type, do the returned objects contain same fields/are the fields named identically? are there any particular keys/fields that would need a special handling in either usecase?

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /**
+     * Infers the type of a field from Neptune data.
+     *
+     * @param key The key of the field we are attempting to infer.
+     * @param value A value from the key whose type we are attempting to infer.
+     * @return The Apache Arrow field definition of the inferred key/value.
+     */
+    private static Field getArrowFieldForNeptune(String key, Object value)
```
Comment (fal-bharadwaj) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java:67]: Are we missing support for some fields for SPARQL here like DECIMAL etc, comparing it with NeptuneSparqlConnection.java I see some are missing, are they being converted to the ones below somewhere?

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java [hunk 10]
```diff
@@ -174,11 +174,9 @@ else if (queryMode.equals(Constants.QUERY_MODE_CLASS)) {
         final GeneratedRowWriter rowWriter = builder.build();
 
         // get results
-        String strim = recordsRequest.getSchema().getCustomMetadata().get(Constants.SCHEMA_STRIP_URI);
-        boolean trimURI = strim == null ? false : Boolean.parseBoolean(strim);
```
Comment (fal-bharadwaj) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java:178]: Why did we delete this? We no longer need to trim the result?

**Review by fal-bharadwaj — CHANGES_REQUESTED** [review]
Please take a look at the comments, thanks.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java (hunk 5 above)
Comment (VenkatasivareddyTR) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java:55]: > Is that limitation for neptune, federated neptune connector or for QPT feature?

This limitation is only for QPT feature.

> even though both Gremlin and SPARQL return a Map type, do the returned objects contain same fields/are the fields named identically? are there any particular keys/fields that would need a special handling in either usecase?

Even though both Gremlin and SPARQL return a Map type, the returned objects generally do not contain the same fields or identically named fields. The map structure is dynamic and specific to each query result, where keys represent column names and values are the corresponding column values.

We process this dynamic map to prepare the schema on the fly based on the query output. Therefore, no special handling is required for particular keys/fields in either the Gremlin or SPARQL use case, as our schema preparation logic is designed to be adaptive.

File: athena-neptune/src/test/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtilsTest.java (hunk 7 above)
Comment (VenkatasivareddyTR) [reply athena-neptune/src/test/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtilsTest.java]: We have added more tests to cover positive and negative use-cases including edge cases.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneGremlinQueryPassthrough.java (hunk 8 above)
Comment (VenkatasivareddyTR) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneGremlinQueryPassthrough.java:85]: Added validation for Unsupported gremlin query format. Null checks and empty checks are already handled in parent class.
We have removed Unsupported gremlin query format validation check from NeptuneMetadataHandler as we are handling here.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java (hunk 9 above)
Comment (VenkatasivareddyTR) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java:67]: No, we are not missing. We are converting to varchar, as we don't have support for process decimal values in SparqlRowWriter.writeRowTemplate()/VertexRowWriter.writeRowTemplate() functions. Even if we get decimal value in NeptuneSparqlConnection.java class we are converting to varchar. Let us know if you have any doubts we can discuss further.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java (hunk 10 above)
Comment (VenkatasivareddyTR) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java:178]: We never used 'trimURI in NeptuneSparqlConnection.next() function, so we cleaned it.

**Comment by VenkatasivareddyTR** [comment]
> Please take a look at the comments, thanks. Some automated checks also failed, please review them as well.

Thank you @fal-bharadwaj for reviewing, we have addressed your review comments, please let us know if you have any other review comments. Fixed build issue after rebasing.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java (hunk 10 above)
Comment (fal-bharadwaj) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java:178]: Not sure I understand, looks like we did use it in line 181 below and then removed it from being used in neptuneConnection.next()?

Comment (VenkatasivareddyTR) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java:178]: In 181 we are passing to neptuneConnection.next() function, however we never used inside next() function. Inside next() function we are using class level variable. Please see below screenshot for your reference.

image

**Review by fal-bharadwaj — APPROVED** [review]
looks good, thanks for addressing the comments!
